src/lib/homestore/heap_chunk_selector.cpp

```diff
@@ -0,0 +1,110 @@
+#include "heap_chunk_selector.h"
+
+#include <execution>
+#include <algorithm>
+#include <sisl/logging/logging.h>
+
+namespace homeobject {
+
+//https://github.com/facebook/folly/blob/61c11d77eb9a8bdc60f673017fccfbe900125cb6/folly/AtomicHashMap.h#L42
+//AtomichashMap has a max size limit of ~18x initial size.
+//we support max of 64k chunks for now, so 64*1024/18 + 1 = 3641, we use 4000 for sufficiently initial size 
+#define DEFAULT_ATOMICHASHMAP_INITIAL_SIZE 4000
+
+HeapChunkSelector::HeapChunkSelector() : 
+m_pdev_heap_map(PdevHeapMap(DEFAULT_ATOMICHASHMAP_INITIAL_SIZE)), 
+m_pdev_avalable_blk_map(PdevAvalableBlkMap(DEFAULT_ATOMICHASHMAP_INITIAL_SIZE)), 
+m_chunks(DEFAULT_ATOMICHASHMAP_INITIAL_SIZE) {}
+
+HeapChunkSelector::HeapChunkSelector(const uint32_t& pdev_heap_map_initial_size, 
+const uint32_t& pdev_avalable_blk_map_initial_size, const uint32_t& chunk_initial_num) : 
+m_pdev_heap_map(PdevHeapMap(pdev_heap_map_initial_size)), 
+m_pdev_avalable_blk_map(PdevAvalableBlkMap(pdev_avalable_blk_map_initial_size)),
+m_chunks(ChunkMap(chunk_initial_num)) {}
+
+
+//add_chunk may be called in Homeobject(when sealing a shard) or in Homestore(when adding chunks to vdev) concurently
+void HeapChunkSelector::add_chunk(csharedChunk& chunk) {
+    VChunk vchunk(chunk);
+    auto&& pdevID = vchunk.get_pdev_id();
+    auto&& [it, _] = m_pdev_heap_map.emplace(pdevID, std::make_shared<std::pair<std::mutex, VChunkHeap>>());
+    const auto& available_blk_num = vchunk.available_blks();
+    auto&& [ignore, ok] = m_pdev_avalable_blk_map.emplace(pdevID, available_blk_num);
```

Comment:
maybe s/ok/inserted for better readability

Reply:
I like `auto [it, happened]` since it reads like english.

Reply:
change to happend to keep the same style with current code